homeassistant/components/media_player/directv.py

```diff
@@ -64,29 +64,57 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         name = 'DirecTV_{}'.format(discovery_info.get('serial', ''))
 
         # Attempt to discover additional RVU units
+        _LOGGER.debug("Doing discovery of DirecTV devices on %s", host)
+
+        from DirectPy import DIRECTV
+        dtv = DIRECTV(host, DEFAULT_PORT)
         try:
-            resp = requests.get(
-                'http://%s:%d/info/getLocations' % (host, DEFAULT_PORT)).json()
-            if "locations" in resp:
-                for loc in resp["locations"]:
-                    if("locationName" in loc and "clientAddr" in loc
-                       and loc["clientAddr"] not in known_devices):
-                        hosts.append([str.title(loc["locationName"]), host,
-                                      DEFAULT_PORT, loc["clientAddr"]])
-
-        except requests.exceptions.RequestException:
+            resp = dtv.get_locations()
+        except requests.exceptions.RequestException as ex:
             # Bail out and just go forward with uPnP data
-            if DEFAULT_DEVICE not in known_devices:
-                hosts.append([name, host, DEFAULT_PORT, DEFAULT_DEVICE])
+            # Make sure that this device is not already configured
+            # Comparing based on host (IP) and clientAddr.
+            _LOGGER.debug("Request exception %s trying to get locations", ex)
+            resp = {
+                'locations': [{
+                    'locationName': name,
+                    'clientAddr': DEFAULT_DEVICE
+                }]
+            }
+
+        _LOGGER.debug("Known devices: %s", known_devices)
+        for loc in resp.get("locations") or []:
+            if "locationName" not in loc or "clientAddr" not in loc:
+                continue
+
+            # Make sure that this device is not already configured
+            # Comparing based on host (IP) and clientAddr.
+            device_unknown = True
+            for device in known_devices:
+                if host in device and loc["clientAddr"] in device:
```

Comment:
Instead of storing a list of four items `[name, host, port, client_address]` in `known_devices`, just store a tuple of `(host, client_address)`. Tuples are hashable so then we can use a single for loop to check membership by creating the tuple to search for from `host` and `client_address`.

We should avoid nesting for loops when possible.

```py
new_device = (host, client_address)
if new_device in known_devices:
   ...
```